x/transferrouter/middlewares/ibc.go

```diff
@@ -0,0 +1,297 @@
+package middlewares
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cometbft/cometbft/crypto/tmhash"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	callbacktypes "github.com/cosmos/ibc-go/v10/modules/apps/callbacks/types"
+	transfertypes "github.com/cosmos/ibc-go/v10/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v10/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v10/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v10/modules/core/exported"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/sagaxyz/saga-sdk/x/transferrouter/keeper"
+	"github.com/sagaxyz/saga-sdk/x/transferrouter/types"
+	"github.com/sagaxyz/saga-sdk/x/transferrouter/utils"
+)
+
+var _ porttypes.IBCModule = IBCMiddleware{}
+
+type IBCMiddleware struct {
+	app            IBCModuleWithUnmarshaler
+	k              keeper.Keeper
+	maxCallbackGas uint64
+}
+
+type IBCModuleWithUnmarshaler interface {
+	porttypes.IBCModule
+	porttypes.PacketDataUnmarshaler
+}
+
+func NewIBCMiddleware(app porttypes.IBCModule, k keeper.Keeper) IBCMiddleware {
+	// check if app implements IBCModuleWithUnmarshaler
+	if _, ok := app.(IBCModuleWithUnmarshaler); !ok {
+		panic("app does not implement IBCModuleWithUnmarshaler")
+	}
+
+	return IBCMiddleware{
+		app: app.(IBCModuleWithUnmarshaler),
+		k:   k,
+	}
+}
+
+// OnAcknowledgementPacket implements types.IBCModule.
+func (i IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	channelVersion string,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	err := i.addSrcCallbackToQueue(ctx, packet, acknowledgement, false)
+	if err != nil {
+		i.k.Logger(ctx).Error("failed to add src callback to queue on acknowledgement packet", "error", err)
+	}
+
+	return i.app.OnAcknowledgementPacket(ctx, channelVersion, packet, acknowledgement, relayer)
+}
+
+// OnTimeoutPacket implements types.IBCModule.
+func (i IBCMiddleware) OnTimeoutPacket(
+	ctx sdk.Context,
+	channelVersion string,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) error {
+	err := i.addSrcCallbackToQueue(ctx, packet, nil, true)
+	if err != nil {
+		i.k.Logger(ctx).Error("failed to add src callback to queue on timeout packet", "error", err)
+	}
+	return i.app.OnTimeoutPacket(ctx, channelVersion, packet, relayer)
+}
+
+// OnRecvPacket implements types.IBCModule.
+func (i IBCMiddleware) OnRecvPacket(ctx sdk.Context, channelVersion string, packet channeltypes.Packet, relayer sdk.AccAddress) exported.Acknowledgement {
```

Comment:
A bit too large with many large blocks, can it be broken down in smaller pieces to make it easier to read?